neuraxle/base.py

```diff
@@ -766,10 +780,39 @@ def update_hyperparams(self, hyperparams: HyperparameterSamples) -> 'BaseStep':
             :func:`~BaseStep.update_hyperparams`,
             :class:`~neuraxle.hyperparams.space.HyperparameterSamples`
         """
-        self.hyperparams.update(hyperparams)
-        self.hyperparams = HyperparameterSamples(self.hyperparams).to_flat()
+        self._update_hyperparams(hyperparams=hyperparams, root=True)
         return self
 
+    def _update_hyperparams(self, hyperparams: HyperparameterSamples, root=False) -> 'BaseStep':
+        hyperparams = HyperparameterSamples(hyperparams).to_flat()
+
+        step_hyperparams, remainders = self._create_recursive_step_values_and_remainders(hyperparams, root)
+        self.hyperparams.update(step_hyperparams)
+
+        self.apply(method_name='_update_hyperparams', hyperparams=remainders, children_only=True)
+
+        return self
+
+    def _create_recursive_step_values_and_remainders(self, hyperparams, root=False):
```

Comment:
This method looks like it should not be in the base step. The base step is not aware of any recursion yet: only the MetaStepMixin and the TruncableSteps are aware of the possibility of a recursion. SRP, and ISP seems broken here: 

> "The interface-segregation principle (ISP) states that no client should be forced to depend on methods it does not use."
- https://hackernoon.com/interface-segregation-principle-bdf3f94f1d11

Reply:
This remark is important here.